### PR TITLE
Add email confirmation to new user accounts

### PR DESCRIPTION
**Pull request by TylerRockwell**
### Why?
https://github.com/smashingboxes/apprenticeship/tree/master/Curriculum/Backend/bookstore
Because we don't want bots on our site.
Because we need valid emails for our future marketing attempts

### What Changed?
Added `:confirmable` to User model
Added migration to add required fields for email confirmation
Configured Action Mailer to send the email
Added integration test for email confirmation

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -38,5 +38,5 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
-  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+  config.action_mailer.default_url_options = { host: 'localhost:3000' }
```
Comment (dkniffin) [inline config/environments/development.rb]: Huh. Did it not work without this change?

**Comment by dkniffin** [comment]
The code looks good, but it looks like you're missing tests for it.

**Comment by TylerRockwell** [comment]
Thanks, @dkniffin ! I'll get some tests put in.

File: config/environments/development.rb (hunk 1 above)
Comment (TylerRockwell) [inline config/environments/development.rb]: It worked, but this was shorter and still just as clear.

Comment (dkniffin) [inline config/environments/development.rb]: :+1:

**Comment by TylerRockwell** [comment]
@dkniffin I added an integration test for the email confirmation, if you'd like to take a look. If not,  I can post it back up in code-reviews.

**Comment by dkniffin** [comment]
I'll take a look

File: features/scenarios/book-index-page.feature [hunk 2]
```diff
@@ -1 +0,0 @@
-
```
Comment (dkniffin) [inline features/scenarios/book-index-page.feature]: Looks like you have an empty file. Either delete it or add something to it.

File: features/step_definitions/account_creation_steps.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
+Given(/^I do not have an account on the site$/) do
+  User.find_by_email('my_email@example.com') == nil
```
Comment (dkniffin) [inline features/step_definitions/account_creation_steps.rb]: 1. Don't use `find_by_<attribute>`. Use `find_by(<attribute>: ...)` instead
2. Here you're testing that you don't have an account. Instead, do `User.find_by(email: 'my_email@example.com').delete_all`

File: features/scenarios/account_creation.feature [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    When I visit the site root path
+    Then I am presented with a login page
+    When I click the "Sign up" link
+    And I enter my email address
+    And I enter a password with correct confirmation
+    And I click the "Sign up" button
+    Then I am told to check my email for a confirmation link
+    And I am sent a confirmation email
+    When I visit the link in that email
+    Then My email address becomes confirmed
```
Comment (dkniffin) [inline features/scenarios/account_creation.feature]: I would reword this "My account becomes confirmed", but it's alright as-is

File: features/step_definitions/account_creation_steps.rb (hunk 3 above)
Comment (TylerRockwell) [inline features/step_definitions/account_creation_steps.rb]: That makes more sense. I'll change it.

**Comment by dkniffin** [comment]
This looks great! When I wrote the comment, I meant to do a unit test (rspec), but this probably tests it better. :+1:

**Comment by TylerRockwell** [comment]
Cool. I have to write these feature tests for the assignment anyway, and I agree that this is probably a better test.

**Comment by TylerRockwell** [comment]
Thanks again, @dkniffin !
